cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i

```diff
@@ -378,56 +415,66 @@ template <typename algorithmFPType, CpuType cpu>
 {
-    algorithmFPType xi = this->_aResponse[aIdx[0]].val;
     ImpurityData left;
-    left.var  = 0;
-    left.mean = xi;
-
     ImpurityData right;
-    calcPrevImpurity<algorithmFPType, cpu>(curImpurity.var * algorithmFPType(n), curImpurity.mean, right.var, right.mean, xi, n);
-
+    algorithmFPType xi = this->_aResponse[aIdx[0]].val;
+    left.var           = 0;
+    left.mean          = xi;
+    IndexType iBest    = -1;
+    algorithmFPType vBest;
+    auto aResponse = this->_aResponse.get();
+    auto aWeights  = this->_aWeights.get();
+    auto weights0  = aWeights[aIdx[0]].val;
+    auto weights   = aWeights[aIdx[n - 1]].val;
+    calcPrevImpurity<algorithmFPType, cpu>(curImpurity.var * totalWeights, curImpurity.mean, right.var, right.mean, xi, totalWeights, weights);
 #ifdef DEBUG_CHECK_IMPURITY
     checkImpurityInternal(aIdx + 1, n - 1, right);
 #endif
-    algorithmFPType vBest = split.impurityDecrease < 0 ? daal::services::internal::MaxVal<algorithmFPType>::get() :
-                                                         (curImpurity.var - split.impurityDecrease) * algorithmFPType(n);
-    IndexType iBest = -1;
+
+    vBest = split.impurityDecrease < 0 ? daal::services::internal::MaxVal<algorithmFPType>::get() :
+                                         (curImpurity.var - split.impurityDecrease) * totalWeights;
+    algorithmFPType leftWeights = 0.;
     for (size_t i = 1; i < (n - nMinSplitPart + 1); ++i)
     {
+        weights = aWeights[aIdx[i]].val;
+        leftWeights += weights;
         const bool bSameFeaturePrev(featureVal[i] <= featureVal[i - 1] + accuracy);
-        if (!(bSameFeaturePrev || (i < nMinSplitPart) || (i < minWeightLeaf) || ((n - i) < minWeightLeaf)))
+
+        if (!(bSameFeaturePrev || (i < nMinSplitPart) || (leftWeights < minWeightLeaf) || ((totalWeights - leftWeights) < minWeightLeaf)))
         {
             //can make a split
             //nLeft == i, nRight == n - i
             const algorithmFPType v = left.var + right.var;
             if (v < vBest)
             {
-                vBest           = v;
-                split.left.var  = left.var;
-                split.left.mean = left.mean;
-                iBest           = i;
+                vBest             = v;
+                split.left.var    = left.var;
+                split.left.mean   = left.mean;
+                split.leftWeights = leftWeights;
+                iBest             = i;
             }
         }
 
         //update impurity and continue
-        xi                    = this->_aResponse[aIdx[i]].val;
+        xi                    = aResponse[aIdx[i]].val;
         algorithmFPType delta = xi - left.mean;
-        left.mean += delta / algorithmFPType(i + 1);
-        left.var += delta * (xi - left.mean);
+        left.mean += weights * delta / (isPositive<algorithmFPType, cpu>(leftWeights + weights0) ? leftWeights + weights0 : 1.0);
```

Comment:
Constant style inconsistency (vs algorithmFPType leftWeights = 0.; )

Reply:
fixed in all cases